### PR TITLE
Voting endpoints

### PR DESCRIPTION
**Pull request by danielgaldev**
Voting request types:
- PUT /questions/\<uid> 
- PUT /answers/\<uid>
- PUT /comments/\<uid>

Request data: `user_vote: "{up|down|none}"`

Complete with 30 running tests testing one-user and two-user voting. Run with `manage.py test question`

Issue: https://github.com/DevTeamSCH/vikoverflow/issues/10

### CODE REVIEW

File: src/question/views.py [hunk 1]
```diff
@@ -1,3 +1,72 @@
...
+    try:
+        vote = request.data['user_vote']
+    except KeyError:
+        return bad_request
+    if vote == 'up':
+        if downvoters.filter(user=user).count() > 0:
+            downvoters.remove(user_profile)
+            upvoters.add(user_profile)
+        else:
+            upvoters.add(user_profile)
```
Comment (ferencmarkizay) [inline src/question/views.py]: This is redundant, calling `upvoters.add(...)` outside the if is enough.

File: src/question/views.py [hunk 2]
```diff
@@ -1,3 +1,72 @@
...
+    elif vote == 'down':
+        if upvoters.filter(user=user).count() > 0:
+            upvoters.remove(user_profile)
+            downvoters.add(user_profile)
+        else:
+            downvoters.add(user_profile)
```
Comment (ferencmarkizay) [inline src/question/views.py]: This is redundant, calling `downvoters.add(...)` outside the if is enough.

File: src/question/views.py [hunk 3]
```diff
@@ -1,3 +1,72 @@
...
+            downvoters.remove(user_profile)
+    else:
+        return bad_request
+    return HttpResponse(status=status.HTTP_200_OK)
+
+
+class AnswerVoteViewSet(mixins.UpdateModelMixin, viewsets.GenericViewSet):
+    queryset = models.Answer.objects.all()
+    serializer_class = serializers.AnswerSerializer
+
```
Comment (ferencmarkizay) [inline src/question/views.py]: Only authenticated users should be able to use this endpoint. 

`permission_classes = (IsAuthenticated, )`

File: src/question/views.py [hunk 4]
```diff
@@ -1,3 +1,72 @@
...
+    def update(self, request, pk):
+        answer = get_object_or_404(models.Answer, pk=pk)
+        return handle_vote(answer, request)
+
+
+class CommentVoteViewSet(mixins.UpdateModelMixin, viewsets.GenericViewSet):
+    queryset = models.Comment.objects.all()
+    serializer_class = serializers.CommentSerializer
+
```
Comment (ferencmarkizay) [inline src/question/views.py]: Only authenticated users should be able to use this endpoint. 

`permission_classes = (IsAuthenticated, )`

File: src/question/views.py [hunk 5]
```diff
@@ -1,3 +1,72 @@
...
+        return handle_vote(answer, request)
+
+
+class QuestionViewSet(
+    mixins.RetrieveModelMixin,
+    viewsets.GenericViewSet
+):
+    queryset = models.Question.objects.all()
+    serializer_class = serializers.QuestionSerializer
+
```
Comment (ferencmarkizay) [inline src/question/views.py]: Only authenticated users should be able to use this endpoint. 

`permission_classes = (IsAuthenticated, )`

File: src/question/tests.py [hunk 6]
```diff
@@ -1,3 +1,255 @@
...
+        self.client.logout()
+
+        self.assertEqual(self.upvoters.all().count(), 0)
+        self.assertEqual(self.downvoters.all().count(), 0)
+        self.assertEqual(self.upvoters.filter(user=self.voter).count(), 0)
+        self.assertEqual(self.downvoters.filter(user=self.voter).count(), 0)
+        self.assertEqual(self.upvoters.filter(user=self.voter2).count(), 0)
+        self.assertEqual(self.downvoters.filter(user=self.voter2).count(), 0)
+
+
```
Comment (ferencmarkizay) [inline src/question/tests.py]: Some more tests would be nice to test errors: a user trying to vote without being logged in, vote on a question that doesn't exist, etc.

File: src/question/views.py (hunk 1 above)
Comment (danielgaldev) [reply src/question/views.py]: Fixed in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

File: src/question/views.py (hunk 2 above)
Comment (danielgaldev) [reply src/question/views.py]: Fixed in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

File: src/question/views.py (hunk 3 above)
Comment (danielgaldev) [reply src/question/views.py]: Fixed in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

File: src/question/views.py (hunk 4 above)
Comment (danielgaldev) [reply src/question/views.py]: Fixed in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

File: src/question/views.py (hunk 5 above)
Comment (danielgaldev) [reply src/question/views.py]: Fixed in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

File: src/question/tests.py (hunk 6 above)
Comment (danielgaldev) [reply src/question/tests.py]: Added tests to test non-existent question and no login in https://github.com/DevTeamSCH/vikoverflow-backend/pull/12/commits/01b536a456a5d593cff71ce02dcf44d2a7b10a32

**Comment by danielgaldev** [comment]
Voting endpoints as of now:
- PUT /questions/\<uid>/vote/
- PUT /answers/\<uid>/vote/
- PUT /comments/\<uid>/vote/

**Pull request merged**